light-client/relayer/src/relayer.rs

```diff
@@ -0,0 +1,286 @@
+            next_block_index,
+            ledger_db,
+            watcher_db,
+            sender,
+            verifier,
+            stop_requested,
+            shared_state,
+            logger,
+        };
+        thread.run();
+    }
+
+    pub fn run(mut self) {
+        // Poll ledger for data to relay
+        log::info!(self.logger, "Relayer thread started.");
+        loop {
+            self.shared_state().current_block_index = self.next_block_index;
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Relayer thread stop requested.");
+                break;
+            }
+            match self.ledger_db.get_block_data(self.next_block_index) {
+                Err(LedgerError::NotFound) => std::thread::sleep(Self::POLLING_FREQUENCY),
+                Err(e) => {
+                    log::error!(
+                        self.logger,
+                        "Unexpected error when checking for block data {}: {:?}",
+                        self.next_block_index,
+                        e
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                }
+                Ok(block_data) => {
+                    if let Err(err) = self.process_block(&block_data) {
+                        log::error!(
+                            self.logger,
+                            "When processing block {}: {:?}",
+                            self.next_block_index,
+                            err
+                        );
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    } else {
+                        self.next_block_index += 1;
+                    }
+                }
+            }
+        }
+    }
+
+    fn process_block(&mut self, block_data: &BlockData) -> Result<(), Error> {
+        let tracer = tracer!();
+
+        let mut span =
+            block_span_builder(&tracer, "poll_block", self.next_block_index).start(&tracer);
+
+        span.set_attribute(TELEMETRY_BLOCK_INDEX_KEY.i64(self.next_block_index as i64));
+
+        let relevant_burns = Self::check_for_relevant_burns(&block_data.contents().outputs);
+        if relevant_burns.is_empty() {
+            return Ok(());
+        }
+
+        // Try to get signatures from the watcher
+        tracer.in_span("loop_for_signatures", |_cx| loop {
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Relayer thread stop requested.");
+                return Ok(());
+            }
+            let signatures = self.get_block_signatures(self.next_block_index)?;
+            let burned = BurnTx {
+                block: block_data.block().clone(),
+                block_contents: block_data.contents().clone(),
+                signatures,
+                tx_outs: relevant_burns.clone(),
+            };
+            match self.verifier.verify_burned_tx(burned.clone()) {
```

Comment:
This feels convoluted and will not work once we just relay on the watcher data and not the ledger. The API you are going to want to use is https://github.com/mobilecoinfoundation/mobilecoin/blob/master/light-client/verifier/src/verifier.rs#L121. You get the `BlockData` array straight from the watcher, pass it here, and if the data is valid you will get the information about the block and can then proceed to look at its TxOuts. 
I think this will also help you get rid of the nested loops, since you can just return an error from this function until the verification succeeds, and then the outer loop will retry.
This does have the implication that we'll do a signature check for all blocks, even those that don't have a burn tx, but I think that's perfectly fine.

Reply:
As discussed offline, this will be done as a separate PR.